File: adventofcode/java/year2016/day17/Day17Test.java
package adventofcode.java.year2016.day17;

import java.io.FileReader;
import java.io.IOException;
import java.io.StringReader;

import org.junit.Test;
import org.junit.Assert;

public class Day17Test {
    @Test
    public void part1Example1() {
        var input = new StringReader("ihgpwlah");
        var output = "DDRRRD";
        var result = Day17.part1(input);
        Assert.assertEquals("no error", "", result.error);
        Assert.assertEquals("correct output", output, result.answer);
    }

    @Test
    public void part1Example2() {
        var input = new StringReader("kglvqrro");
        var output = "DDUDRLRRUDRD";
        var result = Day17.part1(input);
        Assert.assertEquals("no error", "", result.error);
        Assert.assertEquals("correct output", output, result.answer);
    }

    @Test
    public void part1Example3() {
        var input = new StringReader("ulqzkmiv");
        var output = "DRURDRUDDLLDLUURRDULRLDUUDDDRR";
        var result = Day17.part1(input);
        Assert.assertEquals("no error", "", result.error);
        Assert.assertEquals("correct output", output, result.answer);
    }

    @Test
    public void part1Actual() throws IOException {
        try (var input = new FileReader("adventofcode/data/year2016/day17/actual.in")) {
            var output = "DDRLRRUDDR";
            var result = Day17.part1(input);
            Assert.assertEquals("no error", "", result.error);
            Assert.assertEquals("correct output", output, result.answer);
        }
    }

    @Test
    public void part2Example1() {
        var input = new StringReader("ihgpwlah");
        var output = "370";
        var result = Day17.part2(input);
        Assert.assertEquals("no error", "", result.error);
        Assert.assertEquals("correct output", output, result.answer);
    }

    @Test
    public void part2Example2() {
        var input = new StringReader("kglvqrro");
        var output = "492";
        var result = Day17.part2(input);
        Assert.assertEquals("no error", "", result.error);
        Assert.assertEquals("correct output", output, result.answer);
    }

    @Test
    public void part2Example3() {
        var input = new StringReader("ulqzkmiv");
        var output = "830";
        var result = Day17.part2(input);
        Assert.assertEquals("no error", "", result.error);
        Assert.assertEquals("correct output", output, result.answer);
    }

    @Test
    public void part2Actual() throws IOException {
        try (var input = new FileReader("adventofcode/data/year2016/day17/actual.in")) {
            var output = "556";
            var result = Day17.part2(input);
            Assert.assertEquals("no error", "", result.error);
            Assert.assertEquals("correct output", output, result.answer);
        }
    }
}

File: adventofcode/cpp/year2019/day17/day17_test.cc
#include "adventofcode/cpp/year2019/day17/day17.h"

#include <fstream>
#include <string>

#include "gtest/gtest.h"

#include "adventofcode/cpp/adventofcode.h"

TEST(Year2019Day17, Part1Actual) {
  std::ifstream input("adventofcode/data/year2019/day17/actual.in");
  std::string output = "3336";
  adventofcode::answer_t a = day17::part1(input);
  EXPECT_EQ("", a.error);
  EXPECT_EQ(output, a.answer);
  input.close();
}

TEST(Year2019Day17, Part2Actual) {
  std::ifstream input("adventofcode/data/year2019/day17/actual.in");
  std::string output = "597517";
  adventofcode::answer_t a = day17::part2(input);
  EXPECT_EQ("", a.error);
  EXPECT_EQ(output, a.answer);
  input.close();
}
